### PR TITLE
core: implement RuntimeMaxDeltaSec directive

### PR DESCRIPTION
**Pull request by bytesuji**
Implements the feature asked for in #20082

### CODE REVIEW

File: src/core/scope.c [hunk 1]
```diff
@@ -36,6 +37,7 @@ static void scope_init(Unit *u) {
         assert(u->load_state == UNIT_STUB);
 
         s->runtime_max_usec = USEC_INFINITY;
+        s->runtime_max_delta_usec = 0;
```
Comment (bluca) [inline src/core/scope.c]: no need to explicitly initialize to 0

File: src/core/scope.c [hunk 2]
```diff
@@ -51,6 +53,19 @@ static void scope_done(Unit *u) {
         s->timer_event_source = sd_event_source_disable_unref(s->timer_event_source);
 }
 
+static int scope_running_timeout(Scope *s) {
+        usec_t timeout, delta = 0;
+
```
Comment (bluca) [inline src/core/scope.c:57]: `assert(s)`

File: src/core/service.c [hunk 3]
```diff
@@ -514,6 +516,19 @@ static void service_remove_fd_store(Service *s, const char *name) {
         }
 }
 
+static int service_running_timeout(Service *s) {
+        usec_t timeout, delta = 0;
+
```
Comment (bluca) [inline src/core/service.c:520]: `assert(s)`

**Comment by bluca** [comment]
regenerate (and fill in) the documentation with `ninja -C builddir update-dbus-docs`

File: src/core/scope.c (hunk 1 above)
Comment (bytesuji) [reply src/core/scope.c]: Also true in `service.c`?

Comment (bluca) [reply src/core/scope.c]: Yes

**Comment by bluca** [comment]
There are merge conflicts, please rebase and force push

File: man/systemd.scope.xml [hunk 4]
```diff
@@ -108,6 +108,15 @@
         active for longer than the specified time it is terminated and put into a failure state. Pass
         <literal>infinity</literal> (the default) to configure no runtime limit.</para></listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><varname>RuntimeMaxDeltaSec=</varname></term>
+
+        <listitem><para>This option modifies <varname>RuntimeMaxSec=</varname> by adding a random delta to the allowed
+        runtime, which will be less than or equal to the specified value. If <varname>RuntimeMaxSec=</varname> isn't
+        specified, this option will be ignored.
```
Comment (poettering) [inline man/systemd.scope.xml]: what is a random delta? evenly distributed additional time between 0 and the specified value? if so, say so.

please model this after RandomizedDelaySec= for timer units, in name, behaviour and description.

i.e. RuntimeRandomizedExtraSec= or so?

File: src/core/scope.c [hunk 5]
```diff
@@ -51,6 +52,21 @@ static void scope_done(Unit *u) {
...
+        assert(s);
+
+        if (s->runtime_max_delta_usec) {
+                delta = random_u64_range((uint64_t) s->runtime_max_delta_usec);
```
Comment (poettering) [inline src/core/scope.c]: no need to cast explicitly

File: src/core/scope.c [hunk 6]
```diff
@@ -51,6 +52,21 @@ static void scope_done(Unit *u) {
...
+                log_unit_debug(UNIT(s), "Adding delta of %d usec to timeout", (int) delta);
```
Comment (poettering) [inline src/core/scope.c]: FORMAT_TIMESPAN() please

File: src/core/scope.c [hunk 7]
```diff
@@ -51,6 +52,21 @@ static void scope_done(Unit *u) {
...
+        }
+
+        timeout = usec_add(UNIT(s)->active_enter_timestamp.monotonic, s->runtime_max_usec);
+
+        return usec_add(timeout, delta);
```
Comment (poettering) [inline src/core/scope.c]: no need for the tempoary variable, just nest it:

```c
return usec_add(usec_add(UNIT(s)->active_enter_timestamp.monotonic, 
                         s->runtime_max_usec),
                delta);
```

File: src/core/service.c [hunk 8]
```diff
@@ -514,6 +515,21 @@ static void service_remove_fd_store(Service *s, const char *name) {
...
+        if (s->runtime_max_delta_usec) {
+                delta = random_u64_range((uint64_t) s->runtime_max_delta_usec);
+                log_unit_debug(UNIT(s), "Adding delta of %d usec to timeout", (int) delta);
+        }
+
+        timeout = usec_add(UNIT(s)->active_enter_timestamp.monotonic, s->runtime_max_usec);
+
+        return usec_add(timeout, delta);
+}
```
Comment (poettering) [inline src/core/service.c:531]: as for the scope

**Comment by poettering** [comment]
looks pretty OK, but see comments

File: man/systemd.scope.xml [hunk 9]
```diff
@@ -108,6 +108,15 @@
         active for longer than the specified time it is terminated and put into a failure state. Pass
         <literal>infinity</literal> (the default) to configure no runtime limit.</para></listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><varname>RuntimeRandomizedExtraSec=</varname></term>
+
+        <listitem><para>This option modifies <varname>RuntimeMaxSec=</varname> by increasing the maximum runtime by an
+        evenly distributed duration between 0 and the specified value (in seconds). If <varname>RuntimeMaxSec=</varname> is
+        unspecified, then this directive will be ignored.
```
Comment (poettering) [inline man/systemd.scope.xml]: should say what the default is. And "will be ignored" is a bit strange to say. Should say: "then this feature will be disabled"

File: src/core/load-fragment-gperf.gperf.in [hunk 10]
```diff
@@ -381,6 +381,7 @@ Service.TimeoutAbortSec,                 config_parse_service_timeout_abort,
 Service.TimeoutStartFailureMode,         config_parse_service_timeout_failure_mode,   0,                                  offsetof(Service, timeout_start_failure_mode)
 Service.TimeoutStopFailureMode,          config_parse_service_timeout_failure_mode,   0,                                  offsetof(Service, timeout_stop_failure_mode)
 Service.RuntimeMaxSec,                   config_parse_sec,                            0,                                  offsetof(Service, runtime_max_usec)
+Service.RuntimeRandomizedExtraSec,              config_parse_sec,                            0,                                  offsetof(Service, runtime_rand_extra_usec)
```
Comment (poettering) [inline src/core/load-fragment-gperf.gperf.in]: please align the table properly

File: src/core/load-fragment-gperf.gperf.in [hunk 11]
```diff
@@ -533,6 +534,7 @@ Path.DirectoryMode,                      config_parse_mode,
 {{ CGROUP_CONTEXT_CONFIG_ITEMS('Scope') }}
 {{ KILL_CONTEXT_CONFIG_ITEMS('Scope') }}
 Scope.RuntimeMaxSec,                     config_parse_sec,                            0,                                  offsetof(Scope, runtime_max_usec)
+Scope.RuntimeRandomizedExtraSec,                config_parse_sec,                            0,                                  offsetof(Scope, runtime_rand_extra_usec)
```
Comment (poettering) [inline src/core/load-fragment-gperf.gperf.in]: ditto

File: src/core/scope.c [hunk 12]
```diff
@@ -51,6 +52,21 @@ static void scope_done(Unit *u) {
         s->timer_event_source = sd_event_source_disable_unref(s->timer_event_source);
 }
 
+static int scope_running_timeout(Scope *s) {
+        usec_t delta = 0;
+
+        assert(s);
+
+        if (s->runtime_rand_extra_usec) {
```
Comment (poettering) [inline src/core/scope.c]: as per coding style: for arithmetic types (such as integers) please do explicit non-zero comparisons. We only use C's downgrade to bool feature for things that are actual bools or boolean-like, and as the one special exception for pointers. But for numeric types, please compare explicitly with `!= 0`

**Comment by poettering** [comment]
thanks! lgtm!

**Comment by poettering** [comment]
CI failures appear unrelated.

**Pull request merged**